### PR TITLE
refactor(hooks): hoist transcript utils to _lib

### PR DESCRIPTION
**Pull request by devseunggwan**
## 개요

7개 훅이 각자 재구현하던 transcript JSONL 스캔 로직을 `hooks/_lib/_transcript.py` 단일 SoT로 수렴하고, 2개 advisory 훅의 중복 상수 `_TRANSCRIPT_SCAN_LINES = 400`을 import로 대체합니다. 2026-06-04 감사가 지목한 근본 원인(이중 SoT drift)과 동일 클래스의 최대 잔존 중복(약 430라인 삭제)입니다.

Closes #643

## 변경 내용

**신설** — `hooks/_lib/_transcript.py` (public API 7종 + 상수 1종):
`load_transcript` / `load_transcript_objs` / `read_transcript_tail` / `get_current_turn` / `extract_last_assistant_text` / `has_tool_in_turn` / `read_last_user_message` / `TRANSCRIPT_SCAN_LINES`

**전환** — 9개 훅:

| 훅 | 대체된 로컬 심볼 |
|---|---|
| `readonly-verify-deferral-gate`, `completion-signal-gate`, `merge-state-claim-gate` | `_load_transcript`, `_get_current_turn`, `_extract_last_assistant_text` (+signal-gate의 `_has_tool_in_turn`) |
| `block-ask-end-option`, `block-manufactured-action-menu` | `_read_last_user_message` |
| `block-gh-issue-create-without-dup-search` | `_read_transcript_tail` |
| `block-sciomc-finding-commit` | `_load_transcript_objs` (+불용 `Path` import 제거) |
| `external-write-falsify-check`, `pre-output-falsification-gate` | `_TRANSCRIPT_SCAN_LINES` 상수 |

**의도된 거동 변화 1건 (hardening)**: canonical 본문은 가장 방어적이던 merge-state-claim-gate 변형의 `isinstance` 가드를 채택. 기존 unguarded 사본은 content 리스트에 bare string이 섞이면 `AttributeError`를 일으켜 `fail_open`이 삼킨 채 turn 스캔 전체가 무력화되는 잠재 결함이 있었음 — 가드 채택으로 해당 케이스에서 스캔이 계속 동작.

**테스트** — `tests/test_transcript.py` 31건: contract(경계/tool_result 스킵/sidechain/크기 한도) + single-source(함수 객체 `is` identity 고정) + 로컬 재정의 재발 방지 grep + fail-open.

## 리뷰

- `oh-my-claudecode:code-reviewer`: **APPROVE** — MEDIUM 1(중복 dict 재조회) + LOW 3(테스트 커버리지) 전부 반영, CRITICAL/HIGH 0
- `praxis:codex-review-wrap` → codex: **클린 패스** ("기존 호출부를 깨뜨리는 명확한 결함 없음"; pytest는 codex 샌드박스 제약으로 미실행 — 로컬 실행으로 대체 검증)

## 검증

- `python3 -m pytest tests/test_transcript.py -q` → **31 passed**
- `ruff check hooks/ tests/test_transcript.py` → All checks passed
- `./scripts/run-tests.sh` → exit=0, **ALL TESTS PASSED** (기존 훅 테스트 전체 포함)
- acceptance grep: `grep -ln "_load_transcript\|def _read_last_user_message\|def _read_transcript_tail\|_TRANSCRIPT_SCAN_LINES" hooks/*/*/impl.py` → **0건**

## 검증 안 된 것

- 라이브 Claude Code 세션에서의 훅 발화는 미실행 (유닛 + 전체 스위트로 대체). 디스패처 in-process import 경로는 single-source 테스트의 importlib 로드로 등가 검증.

## 영향 범위

훅 9개의 transcript 파싱 경로. manifest/패키징 변경 없음 (`check-plugin-manifests.py` OK). 잘못될 경우 영향은 advisory/gate 훅의 오탐·미탐이며 fail-open 계약으로 도구 호출 차단은 발생하지 않음.

Pre-commit verified: ./scripts/run-tests.sh → ALL TESTS PASSED (exit=0); ruff clean; pytest tests/test_transcript.py 31 passed
Caller chain verified: 전환된 9개 impl.py의 호출부 전수 grep로 확인 — 구 심볼 잔존 0건, 함수 객체 identity 테스트로 import 체인 고정

## Summary by CodeRabbit

* **Refactor**
  * Centralized transcript handling into a shared utility to make transcript-based checks more consistent across hooks.

* **Bug Fixes**
  * More tolerant handling of missing, unreadable, or malformed transcripts so safeguards and advisories are less likely to fail or trigger incorrectly.

* **Tests**
  * Added tests covering transcript parsing, boundary cases, and error scenarios to improve reliability.

### CODE REVIEW

File: hooks/_lib/_transcript.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+    Returns None when the file is missing, unreadable, or larger than
+    `max_bytes` — callers treat None as "cannot scan, fail open".
+    Non-JSON lines are skipped, scanning continues.
+    """
+    try:
+        p = Path(path)
+        if not p.is_file() or p.stat().st_size > max_bytes:
+            return None
+        text = p.read_text(encoding="utf-8", errors="replace")
```
[the remark this answers is not in the document]
Comment (devseunggwan) [reply hooks/_lib/_transcript.py]: 수정 반영: 29123a4 — `_read_bounded_text()` 헬퍼로 교체하여 stat() 프리체크 대신 실제 읽은 바이트에 bound를 강제합니다 (`f.read(max_bytes + 1)` 후 overflow 거부). 라이브 세션이 stat과 read 사이에 append하는 race가 더 이상 계약을 깨지 않습니다. pytest 29 passed / run-tests.sh ALL TESTS PASSED.

File: hooks/_lib/_transcript.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+            last_user_idx = i
+        elif isinstance(content, list):
+            non_tool = [
+                b for b in content
+                if isinstance(b, dict) and b.get("type") != "tool_result"
+            ]
+            if non_tool:
+                last_user_idx = i
+    start = 0 if last_user_idx is None else last_user_idx + 1
+    return events[start:]
```
[the remark this answers is not in the document]
Comment (devseunggwan) [reply hooks/_lib/_transcript.py:163]: 거동 변경은 적용하지 않았습니다 — 전제를 실측으로 반박했습니다.

Probe: 실제 세션 transcript 3파일 2,503 이벤트 전수 스캔 → nested `message` 형태 2,503건, flat `{"role","content"}` 형태 0건.

flat 형태는 라이브 transcript에 존재하지 않으므로 turn-scanning 헬퍼(`get_current_turn` 등)에 normalization을 추가하면 pre-hoist 훅 거동과 달라지는 변경만 생기고 실익이 없습니다 (#643은 거동 보존이 원칙). 대신 29123a4에서 모듈 docstring에 이 범위를 명시했습니다: flat 형태는 `read_last_user_message`만 허용, turn-scanning은 의도적으로 nested 전용.
